Add Asynchronous Metric Instruments SDK

ankit-bhargava opened the pull request:
This PR implements the asynchronous instrument classes described in the Metric Instruments API (PR #161). These classes are capable of actually collecting telemetry data unlike no-op implementations included with the API. Instruments employ Aggregators (pending approval in PR #178 and PR #181 ) to combine the updates they receive into meaningful values. 

A test suite which validates each instrument’s ability to record values in single-threaded and concurrent scenarios is included. Functions necessary for the metrics pipeline as a whole such as reference counting are tested as well.

This PR in conjunction with (PR #179) provides support for all metric instruments listed in the spec. This PR will not compile without all the other PRs mentioned in this description. 

Note: We templated instruments to support a wider array of scalar types. Templating prevents us from using out of line declarations as we normally would. As a result, we left all implementation code in the header file. If there are any suggestions to remedy this we will gladly make the change.  Code for the base classes which asynchronous instruments inherit from is in PR #161.

reyang commented on sdk/include/opentelemetry/sdk/metrics/async_instruments.h:
```diff
@@ -0,0 +1,225 @@
+   }
+
+     /*
+      * Activate the intsrument's callback function to record a measurement.  This
```
```suggestion
      * Activate the instrument's callback function to record a measurement.  This
```

reyang commented on sdk/include/opentelemetry/sdk/metrics/async_instruments.h:
```diff
@@ -0,0 +1,225 @@
+#include <map>
+#include <sstream>
+#include <vector>
+#include <memory>
```
Minor: would you sort includes alphabetically unless there is a good reason not to do so?

ankit-bhargava replied:
The format tools reorder the includes but I sorted them where I could.

reyang commented on sdk/include/opentelemetry/sdk/metrics/async_instruments.h:
```diff
@@ -0,0 +1,225 @@
+     }
+
+    // Public mapping from labels (stored as strings) to their respective aggregators
+   std::unordered_map<std::string, std::shared_ptr<Aggregator<T>>> boundAggregators_;
```
There seems to be indentation issue.

reyang commented on sdk/include/opentelemetry/sdk/metrics/async_instruments.h:
```diff
@@ -0,0 +1,225 @@
+       else
+       {
+           if (value < 0){
+               throw std::invalid_argument("Counter instrument updates must be non-negative.");
```
@pyohannes do you think we should check `__EXCEPTIONS` and call `std::terminate`?
I think we shouldn't throw if exception is disabled, however we don't want to kill the process as this is not a critical failure.

ankit-bhargava replied:
I've added a check and call to terminate :)

reyang commented on sdk/include/opentelemetry/sdk/metrics/async_instruments.h:
```diff
@@ -0,0 +1,225 @@
+         return ret;
+     }
+
+    // Public mapping from labels (stored as strings) to their respective aggregators
```
This seems to be another indentation issue, would you rebase and check if the formatter CI would pass?

ankit-bhargava commented:
@reyang Thanks for the review! I rebased and addressed your comments.  Additionally, we discovered a bug related to the way we were dealing with concurrency in our synchronous instruments.  I added fixes in this PR out of convenience (you can see the changes in the `sync_instruments.h` file, but I can remove them and open a new PR if you would prefer that.

ankit-bhargava commented:
@reyang This should be ready to merge unless there are any further concerns
